salt/utils/extend.py

```diff
@@ -0,0 +1,282 @@
+    """
+    templates = []
+    log.debug('Listing contents of {0}'.format(src))
+    for item in os.listdir(src):
+        s = os.path.join(src, item)
+        if os.path.isdir(s):
+            template_path = os.path.join(s, 'template.yml')
+            if os.path.isfile(template_path):
+                try:
+                    templates.append(_get_template(template_path, item))
+                except:
+                    log.error("Could not load template {0}".format(template_path))
+            else:
+                log.debug("Directory does not contain template.yml {0}".format(template_path))
+    return templates
+
+
+def _mergetree(src, dst):
+    """
+    Akin to shutils.copytree but over existing directories, does a recursive merge copy.
+
+    :param src: The source path
+    :type  src: ``str``
+
+    :param dst: The destination path
+    :type  dst: ``str``
+    """
+    for item in os.listdir(src):
+        s = os.path.join(src, item)
+        d = os.path.join(dst, item)
+        if os.path.isdir(s):
+            log.info("Copying folder {0} to {1}".format(s, d))
+            if os.path.exists(d):
+                _mergetree(s, d)
+            else:
+                shutil.copytree(s, d)
+        else:
+            log.info("Copying file {0} to {1}".format(s, d))
+            shutil.copy2(s, d)
+
+
+def _mergetreejinja(src, dst, context):
+    """
+    Merge directory A to directory B, apply Jinja2 templating to both
+    the file/folder names AND to the contents of the files
+
+    :param src: The source path
+    :type  src: ``str``
+
+    :param dst: The destination path
+    :type  dst: ``str``
+
+    :param context: The dictionary to inject into the Jinja template as context
+    :type  context: ``dict``
+    """
+    for item in os.listdir(src):
+        s = os.path.join(src, item)
+        d = os.path.join(dst, item)
+        if os.path.isdir(s):
+            log.info("Copying folder {0} to {1}".format(s, d))
+            if os.path.exists(d):
+                _mergetreejinja(s, d, context)
+            else:
+                os.mkdir(d)
+                _mergetreejinja(s, d, context)
+        else:
+            log.info("Copying file {0} to {1}".format(s, d))
+            d = Template(d).render(context)
+            with open(s, 'r') as source_file:
+                src_contents = source_file.read()
+                dest_contents = Template(src_contents).render(context)
+            with open(d, 'w') as dest_file:
+                dest_file.write(dest_contents)
+
+
+def _prompt_user_variable(var_name, default_value):
+    """
+    Prompt the user to enter the value of a variable
+
+    :param var_name: The question to ask the user
+    :type  var_name: ``str``
+
+    :param default_value: The default value
+    :type  default_value: ``str``
+
+    :rtype: ``str``
+    :returns: the value from the user
+    """
+    return click.prompt(var_name, default=default_value)
+
+
+def _prompt_choice(var_name, options):
+    """
```

Comment:
The salt code style guide prefers single quotes over double quotes. Not a problem, but good for consistency.